### PR TITLE
Revamp of PracticalSupport entry editing

### PR DESCRIPTION
**Pull request by colinxfleming**
I rule and have completed some work on Case Manager that's ready for review!

There's a need for significantly more data with practical support entries; my plan here is to treat them like patients in the call log and have a summary, option to see a note, and option to click in for more detail. To start with that we're going to need to revamp the approach we're taking here to support a lot more data tracking; our two row thing isn't gonna cut it anymore.

Next up  we're adding practical support notes (text field, not for archive), date of purchase (archivable datestamp), and waiver present (boolean, archivable). 

Here we basically change up the display of practical support entries into a string, then move the editing to a modal. 

This pull request makes the following changes:
* Changes practical support display to match call list style patterns

image
image

no issues, but sets up bumps to #3214 and #3171

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

### CODE REVIEW

File: app/helpers/practical_supports_helper.rb [hunk 1]
```diff
@@ -47,4 +47,13 @@ def practical_support_guidance_link
...
   end
+
+  def practical_support_display_text(practical_support)
+    content = [practical_support.support_type]
+    content.push "#{t('common.from')} #{practical_support.source}"
+    content.push "#{t('common.for')} #{number_to_currency(practical_support.amount)}" if practical_support.amount.present?
+    content.push '(confirmed)' if practical_support.confirmed?
+    content.push "#{t('common.on_')} #{practical_support.support_date.display_date}" if practical_support.support_date.present?
+    content.join(' ')
+  end
```
Comment (lomky) [inline app/helpers/practical_supports_helper.rb:58]: This is pretty hardcoded to English grammar, but its probably good enough for now?

The more robust alternative I can think of would be to encode distinct strings for each variation that take variables, allowing each locale to arrange them as it sees fit. See the `guidance_link` entry for a simpler example.

File: app/views/practical_supports/_edit.html.erb [hunk 2]
```diff
@@ -1,65 +1,47 @@
...
-                           step: '0.01' %>
-      </div>
-
-      <div class="col-3">
-        <%= f.date_field :support_date, autocomplete: 'off', hide_label: true %>
-      </div>
-
+<div class="row">
+  <div class="col">
+    <%= bootstrap_form_with model: support,
```
Comment (lomky) [inline app/views/practical_supports/_edit.html.erb:3]: any particular reason we renamed the model here? Just shorter and easier?

File: app/views/practical_supports/_entries.html.erb [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
-  </div>
-  <hr/>
 
-  <% patient.practical_supports.each do |entry| %>
-    <%= render 'practical_supports/edit',
-              patient: patient,
-              practical_support: entry %>
+      <div class="col-1 practical-support-note-preview">
+        <%#= entry.most_recent_note_display_text %>
+      </div>
```
Comment (lomky) [inline app/views/practical_supports/_entries.html.erb:11]: Is this setup for a future ticket?

File: app/helpers/practical_supports_helper.rb (hunk 1 above)
Comment (lomky) [reply app/helpers/practical_supports_helper.rb:58]: Do we want to display the fulfilled or nah? If we're worried about space, I think (C) and (F) are sufficient for this quick-view.
image

File: app/views/practical_supports/_edit.html.erb [hunk 4]
```diff
@@ -1,65 +1,47 @@
...
+                        label_class: 'tooltip-header-checkbox',
+                        data: { 'tooltip_text': practical_support_fulfilled_help_text } %>
+      <% end %>
+    <% end %>
   </div>
-  <hr/>
 </div>
+
+<div class="row">
+  <div class="col">
```
Comment (lomky) [inline app/views/practical_supports/_edit.html.erb:39]: I realize that we don't actually _need_ a `Close` button here, but I would really like one.Right now my muscle memory instinct is to click the button when I'm done, so better it be another way to exit the modal.

File: app/helpers/practical_supports_helper.rb (hunk 1 above)
Comment (lomky) [reply app/helpers/practical_supports_helper.rb:58]: On last thing - this ordering implies that the Confirmed is on the amount, rather than the whole Support. What do you think of moving it to the end?

**Review by lomky — COMMENTED** [review]
This is so cool. A couple questions and a plea for a modal Close button

File: app/helpers/practical_supports_helper.rb (hunk 1 above)
Comment (colinxfleming) [reply app/helpers/practical_supports_helper.rb:58]: I think I'm in favor of the guidance_link strategy if this winds up needing to expand or gets awkward. I'd rather solidify it before we put the time into anything particularly robust, and the spanish is close enough I think with the `de`s and `por`s mostly being clean subs for the froms and fors. 

I think reordering a little bit is a good idea; I'm less worried about space though, I think this frees us up to a lot more by delegating updating out! As a general rule of thumb I don't think we need _every_ unit of data displayed and can iterate a little on that, but I'd believe that this should be rearranged at least a little bit.

File: app/views/practical_supports/_edit.html.erb (hunk 2 above)
Comment (colinxfleming) [reply app/views/practical_supports/_edit.html.erb:3]: pretty much yeah. Less to type!

File: app/views/practical_supports/_edit.html.erb (hunk 4 above)
Comment (colinxfleming) [reply app/views/practical_supports/_edit.html.erb:39]: I think super reasonable

File: app/views/practical_supports/_entries.html.erb (hunk 3 above)
Comment (colinxfleming) [reply app/views/practical_supports/_entries.html.erb:11]: yeah, next up is pretty much 'put notes in here'.

**Review by colinxfleming — COMMENTED** [review]
So three things to do:
* add a close button
* display fulfilled
* slightly reorder the display text
thx @lomky !

File: app/helpers/practical_supports_helper.rb [hunk 5]
```diff
@@ -47,4 +47,15 @@ def practical_support_guidance_link
     link_to t('patient.practical_support.guidance_link', fund: ActsAsTenant.current_tenant.name), url.to_s, target: '_blank'
     # "For guidance on practical support, view the #{link_content}."
   end
+
+  def practical_support_display_text(practical_support)
+    content = []
+    content.push "(#{t('activerecord.attributes.practical_support.confirmed')})" if practical_support.confirmed?
+    content.push "(#{t('activerecord.attributes.practical_support.fulfilled')})" if practical_support.confirmed?
```
Comment (lomky) [inline app/helpers/practical_supports_helper.rb:54]:
```suggestion
    content.push "(#{t('activerecord.attributes.practical_support.fulfilled')})" if practical_support.fulfilled?
```

File: app/views/practical_supports/_edit.html.erb [hunk 6]
```diff
@@ -1,65 +1,51 @@
...
+        local: false do |f| %>
+      <%= f.select :support_type, options_for_select(practical_support_options, support.support_type) %>
+      <%= f.select :source,
+                    options_for_select(practical_support_source_options(support.source),
+                                      support.source) %>
+      <%= f.number_field :amount,
+                          value: number_to_currency(f.object.amount.to_f, unit: '', delimiter: ''),
+                          autocomplete: 'off',
+                          prepend: '$',
+                          step: '0.01' %>
```
Comment (lomky) [inline app/views/practical_supports/_edit.html.erb:15]: A strange behavior, the edit page converts a blank value to `0.00`, and then if you save an unrelated change, it saves that as explicit `0.00`. That's going to be nonsensical for some support types, but I don't immediately know how to avoid this.

**Review by lomky — COMMENTED** [review]
Two bugs! But I think nearly there!

File: app/helpers/practical_supports_helper.rb (hunk 5 above)
Comment (colinxfleming) [reply app/helpers/practical_supports_helper.rb:54]: dorf

**Review by lomky — APPROVED** [review]
gooooo!
